src/Server/SSH/SSHPtyHandlerFactory.h

```diff
@@ -0,0 +1,111 @@
+#include <Server/TCPServerConnectionFactory.h>
+#include <Poco/Util/AbstractConfiguration.h>
+#include "Common/Exception.h"
+#include <Common/logger_useful.h>
+#include "Server/IServer.h"
+#include "Common/SSH/LibSSHLogger.h"
+#include "Server/SSH/SSHBind.h"
+#include "Server/SSH/SSHSession.h"
+
+namespace Poco
+{
+class Logger;
+}
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+class SSHPtyHandlerFactory : public TCPServerConnectionFactory
+{
+private:
+    IServer & server;
+    Poco::Logger * log;
+    ssh::SSHBind bind;
+    unsigned int max_auth_attempts;
+    unsigned int auth_timeout_seconds;
+    unsigned int finish_timeout_seconds;
+    unsigned int event_poll_interval_milliseconds;
+    std::optional<int> read_write_timeout_seconds; // optional here, as libssh has its own defaults
+    std::optional<int> read_write_timeout_micro_seconds;
+
+public:
+    explicit SSHPtyHandlerFactory(
+        IServer & server_, const Poco::Util::AbstractConfiguration & config)
+        : server(server_), log(&Poco::Logger::get("SSHHandlerFactory"))
+    {
+        LOG_INFO(log, "Initializing sshbind");
+        bind.disableDefaultConfig();
+
+
+        String prefix = "ssh.";
+        auto rsa_key = config.getString(prefix + "host_rsa_key", "");
+        auto ecdsa_key = config.getString(prefix + "host_ecdsa_key", "");
+        auto ed25519_key = config.getString(prefix + "host_ed25519_key", "");
+        max_auth_attempts = config.getUInt("max_auth_attempts", 4);
+        auth_timeout_seconds = config.getUInt("auth_timeout_seconds", 10);
+        finish_timeout_seconds = config.getUInt("finish_timeout_seconds", 5);
+        event_poll_interval_milliseconds = config.getUInt("event_poll_interval_milliseconds", 100);
+        if (config.has("read_write_timeout_seconds"))
+        {
+            read_write_timeout_seconds = config.getInt("read_write_timeout_seconds");
+            if (read_write_timeout_seconds.value() < 0)
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Negative timeout specified");
+        }
+        if (config.has("read_write_timeout_micro_seconds"))
+        {
+
+            read_write_timeout_micro_seconds = config.getInt("read_write_timeout_micro_seconds");
+            if (read_write_timeout_micro_seconds.value() < 0)
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Negative timeout specified");
+        }
+
+        if (event_poll_interval_milliseconds == 0)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Poll interval must be positive");
+        if (auth_timeout_seconds * 1000 < event_poll_interval_milliseconds)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Poll interval exceeds auth timeout");
+        if (finish_timeout_seconds * 1000 < event_poll_interval_milliseconds)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Poll interval exceeds finish timeout");
+
+
+        if (rsa_key.empty() && ecdsa_key.empty() && ed25519_key.empty())
```

Comment:
what if there is more than 1 key defined, is it invalid?

Reply:
_Each server host SHOULD have a host key.  Hosts MAY have multiple
   host keys using multiple different algorithms.  Multiple hosts MAY
   share the same host key._ 

This is stated in [RFC](https://www.rfc-editor.org/rfc/rfc4251#section-4.1)

Reply:
What I meant is, what would multiple calls to `setHostKey` do? Will a call override the previous set?

Reply:
Oh, sorry. Yes it is valid. It will set one key for each type. Then libssh uses that key, which was negotiated during key exchange

Reply:
It is probably worth a documentational comment for the function.

I will then write comments to other functions as well

Reply:
Wrote comments